hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala

```diff
@@ -240,24 +243,46 @@ trait SNDArray extends SType {
   override def _typeWithRequiredness: TypeWithRequiredness = RNDArray(elementType.typeWithRequiredness.setRequired(true).r)
 }
 
+sealed abstract class NDArrayIndex
+case class ScalarIndex(i: Value[Long]) extends NDArrayIndex
+case class SliceIndex(begin: Option[Value[Long]], end: Option[Value[Long]]) extends NDArrayIndex
+case object ColonIndex extends NDArrayIndex
+
 trait SNDArrayValue extends SValue {
   def st: SNDArray
 
   override def get: SNDArrayCode
 
   def loadElement(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder): SCode
 
+  def loadElementAddress(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder): Code[Long]
+
   def shapes(cb: EmitCodeBuilder): IndexedSeq[Value[Long]]
 
   def strides(cb: EmitCodeBuilder): IndexedSeq[Value[Long]]
 
-  def outOfBounds(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder): Code[Boolean]
+  def firstDataAddress(cb: EmitCodeBuilder): Value[Long]
+
+  def outOfBounds(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder): Code[Boolean] = {
+    val shape = this.shapes(cb)
+    val outOfBounds = cb.newLocal[Boolean]("sndarray_out_of_bounds", false)
 
-  def assertInBounds(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder, errorId: Int = -1): Code[Unit]
+    (0 until st.nDims).foreach { dimIndex =>
+      cb.assign(outOfBounds, outOfBounds || (indices(dimIndex) >= shape(dimIndex)))
+    }
+    outOfBounds
+  }
 
-  def sameShape(other: SNDArrayValue, cb: EmitCodeBuilder): Code[Boolean]
+  def assertInBounds(indices: IndexedSeq[Value[Long]], cb: EmitCodeBuilder, errorId: Int): Code[Unit] = {
+    val shape = this.shapes(cb)
+    Code.foreach(0 until st.nDims) { dimIndex =>
+      val eMsg = const("Index ").concat(indices(dimIndex).toS)
+        .concat(s" is out of bounds for axis $dimIndex with size ")
+        .concat(shape(dimIndex).toS)
+      (indices(dimIndex) >= shape(dimIndex)).orEmpty(Code._fatalWithID[Unit](eMsg, errorId))
+    }
+  }
```

Comment:
This one isn't your fault since you just moved this code block, but since we are touching it can we just make this a `Unit` method` and append the `Code.foreach` to the code builder?

Reply:
Oof, good call